format-delimited/src/main/java/io/cdap/plugin/format/delimited/dq/TypeInference.java

```diff
@@ -0,0 +1,206 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.format.delimited.dq;
+
+import java.math.BigInteger;
+import java.util.regex.Pattern;
+/**
+ * Type Interface provides utility functions that allow you to detect the types of data.
+ */
+public class TypeInference {
+  private static final Pattern patternInteger = Pattern.compile("^(\\+|-)?\\d+$");
+
+  private static final Pattern patternDouble = Pattern.compile(
+    "^[-+]?"// Positive/Negative sign
+      + "("// BEGIN Decimal part
+      + "[0-9]+([,\\.][0-9]+)?|"// Alternative I (w/o grouped integer part)
+      + "(" // BEGIN Alternative II (with grouped integer part)
+      + "[0-9]{1,3}" // starting digits
+      + "(" // BEGIN grouped part
+      + "((,[0-9]{3})*"// US integer part
+      + "(\\.[0-9]+)?"// US float part
+      + "|" // OR
+      + "((\\.[0-9]{3})*|([ \u00A0\u2007\u202F][0-9]{3})*)"// EU integer part
+      + "(,[0-9]+)?)"// EU float part
+      + ")"// END grouped part
+      + ")" // END Alternative II
+      + ")" // END Decimal part
+      + "([ ]?[eE][-+]?[0-9]+)?$"); // scientific part
+
+  /**
+   * Detect if the given value is a double type.
+   *
+   * <p>
+   * Note:<br>
+   * 1. This method support only English locale.<br>
+   * e.g. {@code TypeInference.isDouble("3.4")} returns {@code true}.<br>
+   * e.g. {@code TypeInference.isDouble("3,4")} returns {@code false}.<br>
+   * 2. Exponential notation can be detected as a valid double.<br>
+   * e.g. {@code TypeInference.isDouble("1.0E+4")} returns {@code true}.<br>
+   * e.g. {@code TypeInference.isDouble("1.0e-4")} returns {@code true}.<br>
+   * e.g. {@code TypeInference.isDouble("1.0e-04")} returns {@code true}.<br>
+   * 3. Numbers marked with a type is invalid.<br>
+   * e.g. {@code TypeInference.isDouble("3.4d")} returns {@code false}.<br>
+   * e.g. {@code TypeInference.isDouble("123L")} returns {@code false}.<br>
+   * 4. White space is invalid.<br>
+   * e.g. {@code TypeInference.isDouble(" 3.4")} returns {@code false}.<br>
+   * e.g. {@code TypeInference.isDouble("3.4 ")} returns {@code false}.<br>
+   * 5. "." is not obligatory.<br>
+   * e.g. {@code TypeInference.isDouble("100")} returns {@code true}.
+   * <P>
+   *
+   * @param value the value to be detected.
+   * @return true if the value is a double type, false otherwise.
+   */
+  public static boolean isDouble(String value) {
+    if (!isEmpty(value) && patternDouble.matcher(value).matches()) {
```

Comment:
- The data type detection solution was adapted from the existing one in wrangler plugin. 

- The `Double.parseDouble()` would parse stringified integers to double values. Then you'll need further functionality to distinct between double values and integer values. 

- `Boolean.parseBoolean(rawVal)`: will return `true` only when `rawVal="true"`, otherwise it will return `false` for any other value as for example: `"false", "foo", "1"`. Thus, you will again need to check whether `rawVal=="false"` primarily.

- Date and time could come up in various formats. The only way to parse is to match the rawValue to a certain format date/time format, as it is already done.